README.md
docker-LEMP
=============================================================

## Introduction
docker-LEMP is a project to help developpers & devops working on differents web technologies. 

With docker-LEMP you can run a local small server including :
- a web server (nginx)
- a application server (PHP-FPM or PHP-FPM phalcon version)
- a DB server (MySQL)
- a memcached server
- a e-mail server
- some interrestings stats with blackfire.io
- data ready for search with elasticsearch

and also give you production/network administrator a solid base to know how this should be deployed in production environment.

In many aspect you can compare it to wamp, xampp, or any equivalent but based on modern containers LXC technology (Docker).

The tool is totaly compatible with many applications & Framework :
- Simple web site
- Simple PHP application
- Symfony2
- Laravel
- Wordpress
- Drupal
- Phalcon
- Joomla
- and many more ! 

## Under the hood
Under the hood, for people already aware of how docker is working docker-LEMP is based on the following images :

- webofmars/nginx (based on official debian lenny image)
- webofmars/php-fpm (based on official debian lenny image)
- webofmars/php-fpm-phalcon (based on official debian lenny image)
- mysql (official)
- memcached (official)
- djfarrelly/maildev
- blackfire/blackfire
- djfarrelly/maildev
- elastichsearch

[![](https://badge.imagelayers.io/webofmars/nginx:latest.svg)](https://imagelayers.io/?images=webofmars/nginx:latest 'Get your own badge on imagelayers.io') [![](https://badge.imagelayers.io/webofmars/php-fpm:latest.svg)](https://imagelayers.io/?images=webofmars/php-fpm:latest 'Get your own badge on imagelayers.io') [![](https://badge.imagelayers.io/webofmars/php-fpm-phalcon:latest.svg)](https://imagelayers.io/?images=webofmars/php-fpm-phalcon:latest 'Get your own badge on imagelayers.io') [![](https://badge.imagelayers.io/mysql:latest.svg)](https://imagelayers.io/?images=mysql:latest 'Get your own badge on imagelayers.io') [![](https://badge.imagelayers.io/memcached:latest.svg)](https://imagelayers.io/?images=memcached:latest 'Get your own badge on imagelayers.io') [![](https://badge.imagelayers.io/djfarrelly/maildev:latest.svg)](https://imagelayers.io/?images=djfarrelly/maildev:latest 'Get your own badge on imagelayers.io') [![](https://badge.imagelayers.io/blackfire/blackfire:latest.svg)](https://imagelayers.io/?images=blackfire/blackfire:latest 'Get your own badge on imagelayers.io') [![](https://badge.imagelayers.io/elasticsearch:latest.svg)](https://imagelayers.io/?images=elasticsearch:latest 'Get your own badge on imagelayers.io')

## Issues :

- Use github to open & track issues you might enconter



CHANGELOG.md
docker-LEMP changelog
========================================================

v1.16.1 :
- added elasticsearch

v1.15.2 :
- blackfire optimisations

v1.15.1 :
- README update for badges imagelayers
- Optimisations of the ngninx launcher script
- switched to build instead of direct images
- better compatibility for win & osx OS
- added maildev image
- added blackfire image
- added vars script that has to be sourced 1st (the run-\* scripts take care of it)

v1.15:
- NEW : all PHP images now includes memcached support. Both for PHP sessions & users/apps storage

v1.14:
- BUGFIX: coorection d'un bug introduit dans la 1.13 avec le script de lancement de nginx '$(tty) ambigous redirect'

v1.13:
- Optimisation du script de lancement nginx

v1.12:
- BUGFIX: run-LEMP ne tournait pas à cause d'un répertoire vide.

v1.11:
- modification des tailles d'upload dans PHP et nginx
- ajout d'une page d'acceuil
- ajout d'un moyen de vider la cache APCu
- ajout d'un script pour changer les permissions des sites déployés
- fix pour le bug OSX avec MySQL
- ajout de nombreux paquets pour PHP
- Configuration du rewrite pour phalcon
- Activation du microcaching Nginx & PHP

v1.10:
- corrige un problème de montage avec les volumes php5-fpm-phalcon. Ils seront dorénavant montés.

v1.9:
- PHP improvements
- better gitignore file

v1.8:
- phalcon bugfix

v1.7:
- réorganisation des fichiers
- coorection d'un bug lié au dernier commit (chemin relatif)
- coorection du run-LE qui ne marchait plus sans PHP (ce qui n'est pas le but)

v1.6:
- bugfix mineur

v1.5:
- ajout de imagemagick et de son support PHP
- image phalcon basée sur webofmars/php-fpm

v1.4 :
- ajout de PhpMyAdmin

v1.3 :
- ajout du support de Phalcon

v1.2 :
- ajout de composer dans php5-fpm
- ajout de la possibilité de modifier le paramétrage php pour php5-fpm
- reorganisation des sources
- ajout de binaires de lancement



data/nginx/var/www/html/tests/memcache.php
<?php

    $mem  = new Memcached();

    $mem->addServer('memcached',11211);

    if( $mem->add("mystr","this is a memcache test!",3600)){
        echo  'Stored the value in memcached try to refresh now';
    } else {
        echo 'Just get the value from memcache ! mystr is "'.$mem->get("mystr").'"';
    }


data/nginx/var/www/html/tests/sessions.php
<?php

    echo 'Session starting...<br />';
    session_start();
     
    echo 'Session started...<br />';
     
    $_SESSION['favcolor'] = 'green';
    $_SESSION['animal']   = 'cat';
    $_SESSION['time']    = time();
     
    echo 'Your SID is '.session_id().'<br/>';
    echo 'and your favorite color is '. $_SESSION['favcolor'].'<br />';

    echo '<br /><a href="/index.php?' . session_id() . '" data-lbox="1" title="Home">Back to home !</a>';


data/nginx/var/www/html/index.php
<!DOCTYPE HTML>
<html>
    <head>
        <title>docker-LEMP by webofmars - PHP index</title>
        <meta http-equiv="content-type" content="text/html; charset=utf-8" />
        <meta name="description"        content="LEMP Docker dev environment" />
        <meta name="keywords"           content="docker, webofmars, LEMP, dev" />

        <script src="//code.jquery.com/jquery-1.10.1.min.js"></script>
        <script src="//code.jquery.com/jquery-migrate-1.2.1.min.js"></script>
        <link href="//netdna.bootstrapcdn.com/bootstrap/3.0.3/css/bootstrap.min.css" rel="stylesheet">
        <script src="//netdna.bootstrapcdn.com/bootstrap/3.0.3/js/bootstrap.min.js"></script>
    </head>
    <body>
    <div class="container">
        <div class="row text-center">
            <p>
                <img src="//lorempixel.com/400/200/business/">
            </p>
            <h1>This is PHP <?php $version  = explode('.', PHP_VERSION); echo "$version[0].$version[1]"; ?> baby ;-)</h1>
            <p>enjoy this simple content or choose any of the other links bellow.</p>
        </div>
        <div class="row">
            <h1>See also...</h1>
            <p>
                <ul>
                    <li><a href='./index.html'>HTML index</a>
                    <li><a href='./apc.php'>See APCu stats</a>
                    <li><a href='./clear-apc.php'>Clear APCu cache</a>
                    <li><a href='./pma/'>PhpMyAdmin</a>
                    <li><a href='./phpMemcachedAdmin/'>Memcached Admin</a>
                    <li><a href='#' onclick='window.location.href = "//" + location.hostname + ":1080/"'>maildev</a>
                </ul>
            <p>
            <h1>Your projects</h1>
            <p>
                <ul>
                    <?php
                        $filelist = array();
                        if ($handle = opendir(".")) {
                            while ($entry = readdir($handle)) {
                                if ($entry != 'pma' && $entry != 'apc.php' && $entry != '.' && 
                                      $entry != '..' && $entry != 'clear-apc.php' &&
                                      $entry != 'index.html' && $entry != 'index.php') {
                                    $filelist[] = $entry;
                                }
                                
                            }
                            closedir($handle);
                        }
                        
                        foreach($filelist as $dir) {
                            print("<li><a href=\"./$dir\">$dir</a>\n");
                        }
                    ?>
                </ul>
            </p>
            <p>
            <h2>Elasticsearch</h2>
            <ul>
                <li><a href='#' onclick='window.location.href = "//" + location.hostname + ":9200/_plugin/head"'>head</a>
                <li><a href='#' onclick='window.location.href = "//" + location.hostname + ":9200/_plugin/hq"'>HQ</a>
                <li><a href='#' onclick='window.location.href = "//" + location.hostname + ":9200/_plugin/kopf"'>kopf</a>
            </ul>
            </p>
        </div>
    </div>

    <div class="container">
        <div class="jumbotron">
            <h3>Contact Us</h3>
            <p><a href="mailto:contact@webofmars.com" class="btn btn-primary btn-lg" role="button">Contact us !</a></p>
        </div>
    </div>

    <div class="container">
        <div class="row">
            <?php //phpinfo(); ?>
        </div>
    </div>
    </body>
</html>


data/nginx/var/www/html/pma/config.inc.php
<?php
/* vim: set expandtab sw=4 ts=4 sts=4: */
/**
 * phpMyAdmin sample configuration, you can use it as base for
 * manual configuration. For easier setup you can use setup/
 *
 * All directives are explained in documentation in the doc/ folder
 * or at <http://docs.phpmyadmin.net/>.
 *
 * @package PhpMyAdmin
 */

/**
 * This is needed for cookie based authentication to encrypt password in
 * cookie
 */
$cfg['blowfish_secret'] = '7c5aPr9u7coPlR542D='; /* YOU MUST FILL IN THIS FOR COOKIE AUTH! */

/**
 * Servers configuration
 */
$i = 0;

/**
 * First server
 */
$i++;
/* Authentication type */
$cfg['Servers'][$i]['auth_type'] = 'cookie';
/* Server parameters */
$cfg['Servers'][$i]['host'] = 'db';
$cfg['Servers'][$i]['connect_type'] = 'tcp';
$cfg['Servers'][$i]['compress'] = false;
$cfg['Servers'][$i]['AllowNoPassword'] = false;

/**
 * phpMyAdmin configuration storage settings.
 */

/* User used to manipulate with storage */
// $cfg['Servers'][$i]['controlhost'] = '';
// $cfg['Servers'][$i]['controlport'] = '';
// $cfg['Servers'][$i]['controluser'] = 'pma';
// $cfg['Servers'][$i]['controlpass'] = 'pmapass';

/* Storage database and tables */
// $cfg['Servers'][$i]['pmadb'] = 'phpmyadmin';
// $cfg['Servers'][$i]['bookmarktable'] = 'pma__bookmark';
// $cfg['Servers'][$i]['relation'] = 'pma__relation';
// $cfg['Servers'][$i]['table_info'] = 'pma__table_info';
// $cfg['Servers'][$i]['table_coords'] = 'pma__table_coords';
// $cfg['Servers'][$i]['pdf_pages'] = 'pma__pdf_pages';
// $cfg['Servers'][$i]['column_info'] = 'pma__column_info';
// $cfg['Servers'][$i]['history'] = 'pma__history';
// $cfg['Servers'][$i]['table_uiprefs'] = 'pma__table_uiprefs';
// $cfg['Servers'][$i]['tracking'] = 'pma__tracking';
// $cfg['Servers'][$i]['userconfig'] = 'pma__userconfig';
// $cfg['Servers'][$i]['recent'] = 'pma__recent';
// $cfg['Servers'][$i]['favorite'] = 'pma__favorite';
// $cfg['Servers'][$i]['users'] = 'pma__users';
// $cfg['Servers'][$i]['usergroups'] = 'pma__usergroups';
// $cfg['Servers'][$i]['navigationhiding'] = 'pma__navigationhiding';
// $cfg['Servers'][$i]['savedsearches'] = 'pma__savedsearches';
// $cfg['Servers'][$i]['central_columns'] = 'pma__central_columns';
// $cfg['Servers'][$i]['designer_settings'] = 'pma__designer_settings';
// $cfg['Servers'][$i]['export_templates'] = 'pma__export_templates';
/* Contrib / Swekey authentication */
// $cfg['Servers'][$i]['auth_swekey_config'] = '/etc/swekey-pma.conf';

/**
 * End of servers configuration
 */

/**
 * Directories for saving/loading files from server
 */
$cfg['UploadDir'] = '';
$cfg['SaveDir'] = '';

/**
 * Whether to display icons or text or both icons and text in table row
 * action segment. Value can be either of 'icons', 'text' or 'both'.
 * default = 'both'
 */
//$cfg['RowActionType'] = 'icons';

/**
 * Defines whether a user should be displayed a "show all (records)"
 * button in browse mode or not.
 * default = false
 */
//$cfg['ShowAll'] = true;

/**
 * Number of rows displayed when browsing a result set. If the result
 * set contains more rows, "Previous" and "Next".
 * Possible values: 25, 50, 100, 250, 500
 * default = 25
 */
//$cfg['MaxRows'] = 50;

/**
 * Disallow editing of binary fields
 * valid values are:
 *   false    allow editing
 *   'blob'   allow editing except for BLOB fields
 *   'noblob' disallow editing except for BLOB fields
 *   'all'    disallow editing
 * default = 'blob'
 */
//$cfg['ProtectBinary'] = false;

/**
 * Default language to use, if not browser-defined or user-defined
 * (you find all languages in the locale folder)
 * uncomment the desired line:
 * default = 'en'
 */
//$cfg['DefaultLang'] = 'en';
//$cfg['DefaultLang'] = 'de';

/**
 * How many columns should be used for table display of a database?
 * (a value larger than 1 results in some information being hidden)
 * default = 1
 */
//$cfg['PropertiesNumColumns'] = 2;

/**
 * Set to true if you want DB-based query history.If false, this utilizes
 * JS-routines to display query history (lost by window close)
 *
 * This requires configuration storage enabled, see above.
 * default = false
 */
//$cfg['QueryHistoryDB'] = true;

/**
 * When using DB-based query history, how many entries should be kept?
 * default = 25
 */
//$cfg['QueryHistoryMax'] = 100;

/**
 * Whether or not to query the user before sending the error report to
 * the phpMyAdmin team when a JavaScript error occurs
 *
 * Available options
 * ('ask' | 'always' | 'never')
 * default = 'ask'
 */
//$cfg['SendErrorReports'] = 'always';

/**
 * You can find more configuration options in the documentation
 * in the doc/ folder or at <http://docs.phpmyadmin.net/>.
 */


data/nginx/var/www/html/clear-apc.php
<!DOCTYPE HTML>
<html>
    <head>
        <title>docker-LEMP by webofmars - clear-apc</title>
        <meta http-equiv="content-type" content="text/html; charset=utf-8" />
        <meta name="description"        content="LEMP Docker dev environment" />
        <meta name="keywords"           content="docker, webofmars, LEMP, dev" />

        <link href="http://netdna.bootstrapcdn.com/bootstrap/3.0.3/css/bootstrap.min.css" rel="stylesheet">

        <script src="http://netdna.bootstrapcdn.com/bootstrap/3.0.3/js/bootstrap.min.js"></script>
        <script src="http://code.jquery.com/jquery-1.10.1.min.js"></script>
        <script src="http://code.jquery.com/jquery-migrate-1.2.1.min.js"></script>
    </head>
    <body>
    <div class="container">
        <div class="row text-center">
            <p>
                <img src="http://lorempixel.com/400/200/business/">
            </p>
            <h1>clear-APC page</h1>
            <p>Results of the clearing : <?php

if (!isset($_SERVER['GATEWAY_INTERFACE'])) {

    $opts = getopt('h', [ 'port:', 'sock:', 'help' ]);
    if (count($opts) === 0) {
        $opts['port'] = 9000;
    } elseif (isset($opts['h']) || isset($opts['help'])) {
        fprintf(STDOUT, 'clear-apc.php 1.0'.PHP_EOL.'Usage: clear-apc.php --port=port|--sock=path'.PHP_EOL);
        exit(1);
    }

    if (isset($opts['sock'])) {
        $host = $opts['sock'];
        $port = null;
    } else {
        $host = '127.0.0.1';
        $port = $opts['port'];
    }

    $headers = [
        'GATEWAY_INTERFACE' => 'FastCGI/1.0',
        'SCRIPT_FILENAME' => realpath($_SERVER['PHP_SELF']),
        'REQUEST_METHOD' => 'POST',
        'REQUEST_URI' => '/'
    ];

    try {
        $client = new FastCgiClient($host, $port);
        $client->request($headers, '');
        $response = $client->response();
        if ($response['statusCode'] !== 200) {
            fprintf(STDERR, $response['stderr'].PHP_EOL);
            exit(1);
        } else {
            fprintf(STDOUT, $response['body'].PHP_EOL);
        }
    } catch (CommunicationException $ex) {
        fprintf(STDERR, $ex->getMessage().PHP_EOL);
        exit(1);
    }
} else {
    $return = apc_clear_cache('opcode');
    echo var_export($return, true) . PHP_EOL;
}

/**
 * Note : Code is released under the GNU LGPL
 *
 * Please do not change the header of this file
 *
 * This library is free software; you can redistribute it and/or modify it under the terms of the GNU
 * Lesser General Public License as published by the Free Software Foundation; either version 2 of
 * the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * See the GNU Lesser General Public License for more details.
 */


/**
 * Handles communication with a FastCGI application
 *
 * @author      Pierrick Charron <pierrick@webstart.fr>
 * @author      Daniel Aharon <dan@danielaharon.com>
 * @author      Erik Bernhardson <bernhardsonerik@gmail.com>
 * @version     2.0
 */
class FastCgiClient
{
    const VERSION_1            = 1;

    const BEGIN_REQUEST        = 1;
    const ABORT_REQUEST        = 2;
    const END_REQUEST          = 3;
    const PARAMS               = 4;
    const STDIN                = 5;
    const STDOUT               = 6;
    const STDERR               = 7;
    const DATA                 = 8;
    const GET_VALUES           = 9;
    const GET_VALUES_RESULT    = 10;
    const UNKNOWN_TYPE         = 11;
    const MAXTYPE              = self::UNKNOWN_TYPE;

    const RESPONDER            = 1;
    const AUTHORIZER           = 2;
    const FILTER               = 3;

    const REQUEST_COMPLETE     = 0;
    const CANT_MPX_CONN        = 1;
    const OVERLOADED           = 2;
    const UNKNOWN_ROLE         = 3;

    const MAX_CONNS            = 'MAX_CONNS';
    const MAX_REQS             = 'MAX_REQS';
    const MPXS_CONNS           = 'MPXS_CONNS';

    const HEADER_LEN           = 8;

    /**
     * Socket
     * @var Resource
     */
    private $sock = null;

    /**
     * Host
     * @var String
     */
    private $host = null;

    /**
     * Port
     * @var Integer
     */
    private $port = null;

    /**
     * Unix socket path
     * @var string
     */
    private $socketPath = null;

    /**
     * Keep Alive
     * @var Boolean
     */
    private $keepAlive = false;

    /**
     * A request has been sent.
     * @var Boolean
     */
    private $awaitingResponse = false;

    /**
     * Constructor
     *
     * @param String $host Host of the FastCGI application or path to the FastCGI unix socket
     * @param Integer $port Port of the FastCGI application or null for the FastCGI unix socket
     */
    public function __construct($host, $port = null)
    {
        if ($port !== null) {
            $this->host = $host;
            $this->port = $port;
        } else {
            $this->socketPath = $host;
        }
    }

    /**
     * Destructor
     */
    public function __destruct()
    {
        if ($this->sock) {
            socket_close($this->sock);
        }
    }

    /**
     * Define whether or not the FastCGI application should keep the connection
     * alive at the end of a request
     *
     * @param Boolean $b true if the connection should stay alive, false otherwise
     */
    public function setKeepAlive($b)
    {
        $this->keepAlive = (boolean)$b;
        if (!$this->keepAlive && $this->sock) {
            $this->close();
        }
    }

    /**
     * Get the keep alive status
     *
     * @return Boolean true if the connection should stay alive, false otherwise
     */
    public function getKeepAlive()
    {
        return $this->keepAlive;
    }

    /**
     * Close the fastcgi connection
     */
    public function close()
    {
        if ($this->sock) {
            socket_close($this->sock);
            $this->sock = null;
        }
    }

    /**
     * Create a connection to the FastCGI application
     */
    private function connect()
    {
        if (!$this->sock) {
            if($this->socketPath !== null) {
                $this->sock = @socket_create(AF_UNIX, SOCK_STREAM, 0);
                $address = $this->socketPath;
                $port = 0;
            } else {
                $this->sock = @socket_create(AF_INET, SOCK_STREAM, SOL_TCP);
                $address = $this->host;
                $port = $this->port;
            }
            if (!$this->sock) {
                throw CommunicationException::socketCreate();
            }
            if (false === @socket_connect($this->sock, $address, $port)) {
                throw CommunicationException::socketConnect($this->sock, $address, $port);
            }
        }
    }

    /**
     * Build a FastCGI packet
     *
     * @param Integer $type Type of the packet
     * @param String $content Content of the packet
     * @param Integer $requestId RequestId
     * @return string
     */
    private function buildPacket($type, $content, $requestId = 1)
    {
        $clen = strlen($content);
        return chr(self::VERSION_1)         /* version */
            . chr($type)                    /* type */
            . chr(($requestId >> 8) & 0xFF) /* requestIdB1 */
            . chr($requestId & 0xFF)        /* requestIdB0 */
            . chr(($clen >> 8 ) & 0xFF)     /* contentLengthB1 */
            . chr($clen & 0xFF)             /* contentLengthB0 */
            . chr(0)                        /* paddingLength */
            . chr(0)                        /* reserved */
            . $content;                     /* content */
    }

    /**
     * Build an FastCGI Name value pair
     *
     * @param String $name Name
     * @param String $value Value
     * @return String FastCGI Name value pair
     */
    private function buildNvpair($name, $value)
    {
        $nlen = strlen($name);
        $vlen = strlen($value);
        if ($nlen < 128) {
            /* nameLengthB0 */
            $nvpair = chr($nlen);
        } else {
            /* nameLengthB3 & nameLengthB2 & nameLengthB1 & nameLengthB0 */
            $nvpair = chr(($nlen >> 24) | 0x80) . chr(($nlen >> 16) & 0xFF) . chr(($nlen >> 8) & 0xFF) . chr($nlen & 0xFF);
        }
        if ($vlen < 128) {
            /* valueLengthB0 */
            $nvpair .= chr($vlen);
        } else {
            /* valueLengthB3 & valueLengthB2 & valueLengthB1 & valueLengthB0 */
            $nvpair .= chr(($vlen >> 24) | 0x80) . chr(($vlen >> 16) & 0xFF) . chr(($vlen >> 8) & 0xFF) . chr($vlen & 0xFF);
        }
        /* nameData & valueData */
        return $nvpair . $name . $value;
    }

    /**
     * Read a set of FastCGI Name value pairs
     *
     * @param String $data Data containing the set of FastCGI NVPair
     * @param null $length
     * @return array of NVPair
     */
    private function readNvpair($data, $length = null)
    {
        $array = array();

        if ($length === null) {
            $length = strlen($data);
        }

        $p = 0;

        while ($p != $length) {

            $nlen = ord($data{$p++});
            if ($nlen >= 128) {
                $nlen = ($nlen & 0x7F << 24);
                $nlen |= (ord($data{$p++}) << 16);
                $nlen |= (ord($data{$p++}) << 8);
                $nlen |= (ord($data{$p++}));
            }
            $vlen = ord($data{$p++});
            if ($vlen >= 128) {
                $vlen = ($nlen & 0x7F << 24);
                $vlen |= (ord($data{$p++}) << 16);
                $vlen |= (ord($data{$p++}) << 8);
                $vlen |= (ord($data{$p++}));
            }
            $array[substr($data, $p, $nlen)] = substr($data, $p+$nlen, $vlen);
            $p += ($nlen + $vlen);
        }

        return $array;
    }

    /**
     * Decode a FastCGI Packet
     *
     * @param String $data String containing all the packet
     * @return array
     */
    private function decodePacketHeader($data)
    {
        $ret = array();
        $ret['version']       = ord($data{0});
        $ret['type']          = ord($data{1});
        $ret['requestId']     = (ord($data{2}) << 8) + ord($data{3});
        $ret['contentLength'] = (ord($data{4}) << 8) + ord($data{5});
        $ret['paddingLength'] = ord($data{6});
        $ret['reserved']      = ord($data{7});
        return $ret;
    }

    /**
     * Read a FastCGI Packet
     *
     * @throws CommunicationException
     * @return array
     */
    private function readPacket()
    {
        $packet = @socket_read($this->sock, self::HEADER_LEN);
        if ($packet === false) {
            throw CommunicationException::socketRead($this->sock);
        }

        $resp = $this->decodePacketHeader($packet);

        if ($len = $resp['contentLength'] + $resp['paddingLength']) {
            $content = @socket_read($this->sock, $len);
            if ($content === false) {
                throw CommunicationException::socketRead($this->sock);
            }
            $resp['content'] = substr($content, 0, $resp['contentLength']);
        } else {
            $resp['content'] = '';
        }

        return $resp;
    }

    /**
     * Get Informations on the FastCGI application
     *
     * @param array $requestedInfo information to retrieve
     * @throws CommunicationException|\Exception
     * @return array
     */
    public function getValues(array $requestedInfo)
    {
        try {
            return $this->doGetValues($requestedInfo);
        } catch (CommunicationException $e) {
            $this->close();
            throw $e;
        }
    }

    /**
     * Get Informations on the FastCGI application
     *
     * @param array $requestedInfo information to retrieve
     * @throws CommunicationException
     * @return array
     */
    protected function doGetValues(array $requestedInfo)
    {
        $this->connect();

        $request = '';
        foreach ($requestedInfo as $info) {
            $request .= $this->buildNvpair($info, '');
        }

        if (false === @socket_write($this->sock, $this->buildPacket(self::GET_VALUES, $request, 0))) {
            throw CommunicationException::socketWrite($this->sock);
        }

        $this->awaitingResponse = true;
        $resp = $this->readPacket();
        $this->awaitingResponse = false;

        if ($resp['type'] == self::GET_VALUES_RESULT) {
            return $this->readNvpair($resp['content'], $resp['length']);
        } else {
            throw new CommunicationException('Unexpected response type, expecting GET_VALUES_RESULT');
        }
    }

    /**
     * Execute a request to the FastCGI application
     *
     * @param array $params Array of parameters
     * @param String $stdin Content
     * @throws CommunicationException|\Exception
     */
    public function request(array $params, $stdin)
    {
        try {
            $this->doRequest($params, $stdin);
        } catch (CommunicationException $e) {
            $this->close();
            throw $e;
        }
    }

    /**
     * Execute a request to the FastCGI application
     *
     * @param array $params Array of parameters
     * @param String $stdin Content
     * @throws CommunicationException
     */
    protected function doRequest(array $params, $stdin)
    {
        $this->connect();

        $request = $this->buildPacket(self::BEGIN_REQUEST, chr(0) . chr(self::RESPONDER) . chr((int) $this->keepAlive) . str_repeat(chr(0), 5));

        $paramsRequest = '';
        foreach ($params as $key => $value) {
            $paramsRequest .= $this->buildNvpair($key, $value);
        }
        if ($paramsRequest) {
            $request .= $this->buildPacket(self::PARAMS, $paramsRequest);
        }
        $request .= $this->buildPacket(self::PARAMS, '');

        if ($stdin) {
            $request .= $this->buildPacket(self::STDIN, $stdin);
        }
        $request .= $this->buildPacket(self::STDIN, '');

        // Write the request and break.
        if (false === @socket_write($this->sock, $request)) {
            throw CommunicationException::socketWrite($this->sock);
        }

        $this->awaitingResponse = true;
    }

    /**
     * FCGIClient::formatResponse()
     *
     * Format the response into an array with separate statusCode, headers, body, and error output.
     *
     * @param string $stdout The plain, unformatted response.
     * @param string $stderr The plain, unformatted error output.
     *
     * @throws CommunicationException
     * @return array An array containing the headers and body content.
     */
    private static function formatResponse($stdout, $stderr) {

        // Split the header from the body.  Split on \n\n.
        $doubleCr = strpos($stdout, "\r\n\r\n");
        $rawHeader = substr($stdout, 0, $doubleCr);
        $rawBody = substr($stdout, $doubleCr, strlen($stdout));

        // Format the header.
        $header = array();
        $headerLines = explode("\n", $rawHeader);

        foreach ($headerLines as $line) {
            if (preg_match('/([\w-]+):\s*(.*)$/', $line, $matches)) {
                // ['Content-type'] => 'text/plain'
                $header[strtolower($matches[1])] = trim($matches[2]);
            }
        }
        if (isset($header['status'])) {
            $code = $header['status'];
            if (false !== ($pos = strpos($code, ' '))) {
                $code = substr($code, 0, $pos);
            }
        } else {
            if (isset($header['location'])) {
                $header['status'] = '302 Moved Temporarily';
                $code = '302';
            }  else {
                $header['status'] = '200 OK';
                $code = '200';
            }
        }

        if (false === ctype_digit($code)) {
            throw new CommunicationException("Unrecognizable status code returned from fastcgi: $code");
        }

        return array(
            'statusCode' => (int) $code,
            'headers'    => $header,
            'body'       => trim($rawBody),
            'stderr'     => $stderr,
        );
    }

    /**
     * Collect the response from a FastCGI request.
     *
     * @throws CommunicationException|\Exception
     * @return String Return response.
     */
    public function response()
    {
        try {
            return $this->doResponse();
        } catch (CommunicationException $e) {
            $this->close();
            throw $e;
        }
    }

    /**
     * Collect the response from a FastCGI request.
     *
     * @throws CommunicationException
     * @return String Return response.
     */
    protected function doResponse()
    {
        $stdout = $stderr = '';
        $resp = false;
        while ($this->awaitingResponse) {

            $resp = $this->readPacket();

            // Check for the end of the response.
            if ($resp['type'] == self::END_REQUEST) {
                $this->awaitingResponse = false;
                // Check for response content.
            } elseif ($resp['type'] == self::STDOUT) {
                $stdout .= $resp['content'];
            } elseif ($resp['type'] == self::STDERR) {
                $stderr .= $resp['content'];
            }
        }

        if (!is_array($resp)) {
            throw new CommunicationException("Bad Request");
        }

        switch (ord($resp['content']{4})) {
            case self::CANT_MPX_CONN:
                throw new CommunicationException('This app can\'t multiplex [CANT_MPX_CONN]');
                break;
            case self::OVERLOADED:
                throw new CommunicationException('New request rejected; too busy [OVERLOADED]');
                break;
            case self::UNKNOWN_ROLE:
                throw new CommunicationException('Role value not known [UNKNOWN_ROLE]');
                break;
            case self::REQUEST_COMPLETE:
                return static::formatResponse($stdout, $stderr);
        }
        return '';
    }

    public function __toString()
    {
        if ($this->awaitingResponse) {
            $status = 'waiting for response';
        } elseif ($this->sock) {
            $status = 'ready for request';
        } else {
            $status = 'not connected';
        }

        $address = $this->socketPath;
        if (!$address) {
            $address = "{$this->host}:{$this->port}";
        }

        return "FCGIClient for $address : $status";
    }
}

/**
 * Handles communication with a FastCGI application
 *
 * @author      Erik Bernhardson <bernhardsonerik@gmail.com>
 * @version     2.0
 */
class CommunicationException extends \RuntimeException
{
    static public function socketCreate()
    {
        $err = socket_last_error();
        return new self("Couldn't create socket - $err - ".socket_strerror($err));
    }

    static public function socketConnect($socket, $host, $port)
    {
        if ($port) {
            $host .= ":$port";
        }
        return self::socketError("Failure connecting to $host", $socket);
    }

    static public function socketRead($socket)
    {
        if ($socket === null) {
            return self::requestAborted();
        }

        return self::socketError('Failure reading socket', $socket);
    }

    static public function socketWrite($socket)
    {
        if ($socket === null) {
            return self::requestAborted();
        }

        return self::socketError('Failure writing socket', $socket);
    }

    static public function socketError($message, $socket)
    {
        $err = socket_last_error($socket);
        return new self("$message - $err - ".socket_strerror($err));
    }

    static public function requestAborted()
    {
        return new self("The request was aborted.");
    }
}
?></p></div>
        <div class="row">
            <h1>See also...</h1>
            <p>
                <ul>
                    <li><a href='./index.php'>PHP index</a>
                    <li><a href='./apc.php'>See APCu stats</a>
                    <li><a href='./clear-apc.php'>Clear APCu cache</a>
                    <li><a href='./pma/'>PhpMyAdmin</a>
                    <li><a href='./phpMemcachedAdmin/'>Memcached Admin</a>
                    <li><a href='#' onclick='window.location.href = "//" + location.hostname + ":1080/"'>maildev</a>
                </ul>
            </p>
            <p>
                <h2>Elasticsearch</h2>
                <ul>
                    <li><a href='#' onclick='window.location.href = "//" + location.hostname + ":9200/_plugin/head"'>head</a>
                    <li><a href='#' onclick='window.location.href = "//" + location.hostname + ":9200/_plugin/hq"'>HQ</a>
                    <li><a href='#' onclick='window.location.href = "//" + location.hostname + ":9200/_plugin/kopf"'>kopf</a>
                </ul>
            </p>
        </div>
    </div>

    <div class="container">
        <div class="jumbotron">
            <h3>Contact Us</h3>
            <p><a href="mailto:contact@webofmars.com" class="btn btn-primary btn-lg" role="button">Contact us !</a></p>
        </div>
    </div>
    </body>
</html>
